### PR TITLE
16 add action selection strategies

### PR DESCRIPTION
**Pull request by parissashahabi**

### CODE REVIEW

File: src/neural_bandits/utils/selectors.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
+from abc import ABC, abstractmethod
+
+import torch
+
+
+class AbstractSelector(ABC):
+    """Defines the interface for all bandit action selectors.
+    Given a tensor of scores per action, the selector chooses the best action (i.e. an arm)
```
Comment (Rob2U) [inline src/neural_bandits/utils/selectors.py:8]: nit: I would omit the "best" because we do not really know what the best action is.

File: src/neural_bandits/utils/selectors.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    or the best set of actions (i.e. a super arm in combinatorial bandits). The selector
+    returns a one hot encoded tensor of the chosen actions.
+    """
+
+    @abstractmethod
+    def __call__(self, scores: torch.Tensor) -> torch.Tensor:
+        """Selects a single best action, or a set of actions in the case of combinatorial bandits.
+
+        Args:
+            scores (torch.Tensor): Tensor of shape (batch_size, n_arms).
```
Comment (Rob2U) [inline src/neural_bandits/utils/selectors.py:18]: nit: We should not provide the types in the docstring. I think with the type annotations they are redundant and easy to miss when refactoring a function interface.

File: src/neural_bandits/utils/selectors.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                This may contain a probability distribution per sample (when used for thompson sampling)
```
Comment (Rob2U) [inline src/neural_bandits/utils/selectors.py:19]: nit: standard Thompson Sampling samples the weights (I would just omit the part inside the braces).

File: src/neural_bandits/utils/selectors.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+        Args:
+            scores: Tensor of shape (batch_size, n_arms) containing scores for each action.
+
+        Returns:
+            Tensor of shape (batch_size, n_arms) containing one-hot encoded selected actions.
+        """
+        batch_size, n_arms = scores.shape
+
+        random_vals = torch.rand(batch_size)
```
Comment (Rob2U) [inline src/neural_bandits/utils/selectors.py:72]: I would recommend using a `torch.Generator` here. This way the constructor could take a seed and do something like:
```
self.generator = torch.Generator()
if seed is not None:
    self.generator.manual_seed(seed)
```
(This way we do not change anything in the global Random Generator State)

File: src/neural_bandits/utils/selectors.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        Returns:
+            Tensor of shape (batch_size, n_arms) containing one-hot encoded selected actions,
+            where exactly k entries are 1 per sample.
+        """
+        batch_size, n_arms = scores.shape
+        assert (
+            self.k <= n_arms
+        ), f"k ({self.k}) cannot be larger than number of arms ({n_arms})"
+
+        _, top_k_indices = torch.topk(scores, k=self.k, dim=1)
```
Comment (Rob2U) [inline src/neural_bandits/utils/selectors.py:110]: The [documentation](https://pytorch.org/docs/main/generated/torch.topk.html) states:
> When using torch.topk, the indices of tied elements are not guaranteed to be stable and may vary across different invocations.
This is, however, not the case for `torch.argmax`. Therefore, the TopKSelector needs to break ties manually.

**Comment by philippkolbe** [comment]
looks like we can merge right? @parissashahabi

**Comment by parissashahabi** [comment]
> looks like we can merge right? @parissashahabi

"Review required" :D
(statlog tests are failing again)

**Comment by Rob2U** [comment]
Hey @parissashahabi what was the reason why you decided against a seed + `torch.Generator` for the `EpsilonGreedySelector`?

**Comment by parissashahabi** [comment]
> Hey @parissashahabi what was the reason why you decided against a seed + `torch.Generator` for the `EpsilonGreedySelector`?

I already added it 🤔 What do you mean?
``` python
def __init__(self, epsilon: float = 0.1, seed: Optional[int] = None) -> None:
        """Initialize the epsilon-greedy selector.
        Args:
            epsilon: Exploration probability. Must be between 0 and 1. Defaults to 0.1.
            seed: Random seed for the generator. Defaults to None.
        """
        assert 0 <= epsilon <= 1, "Epsilon must be between 0 and 1"
        self.epsilon = epsilon
        self.generator = torch.Generator()
        if seed is not None:
            self.generator.manual_seed(seed)
```

**Comment by Rob2U** [comment]
My bad sorry. I actually looked at an old version of this PR.

**Review by Rob2U — APPROVED** [review]
Looks very nice 👌
